### PR TITLE
refactor(jobsdb): use a request executor to encapsulate execution mode (concurrent or queued)

### PR DESCRIPTION
**Pull request by atzoum**
## Description of the change

In the previous implementation, each method that wanted to perform a read/write operation was repeating the same logic:

1. Check if read/write should be done through a channel or not.
2. If not, execute the function immediately.
3. If a channel should be used, prepare another struct holding the request type, parameters & response channel, push this struct to the appropriate channel, the worker then would read the request, execute a different function depending on the request type and publish the response to the response channel.

Now instead, we always create a command that holds the function to be executed and this is sent to an executor.
The executor, depending on the runtime configuration of the system will either execute the function synchronously, or push it to the relevant channel and wait until a response is received.

A few advantages introduced with this refactoring:

1. DRY
2. No need to copy function parameters around
3. The executor doesn't need to know any details about the function being called

## Notion Link

https://www.notion.so/rudderstacks/Use-a-request-executor-in-jobsdb-to-encapsulate-execution-mode-concurrent-or-queued-5a26e147d253469aa507ab75a5faf1ff

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: jobsdb/jobsdb.go [hunk 1]
```diff
@@ -1850,7 +1799,7 @@ Next set of functions are for reading/writing jobs and job_status for
 a given dataset. The names should be self explainatory
 */
 func (jd *HandleT) copyJobsDS(txn *sql.Tx, ds dataSetT, jobList []*JobT) error { //When fixing callers make sure error is handled with assertError
-	queryStat := jd.storeTimerStat("store_jobs")
+	queryStat := jd.getTimerStat("store_jobs", nil)
```
Comment (chandumlg) [inline jobsdb/jobsdb.go:1802]: Let's change the state name from store_jobs to copy_jobs?

File: jobsdb/jobsdb.go [hunk 2]
```diff
@@ -3532,26 +3448,11 @@ func (jd *HandleT) store(jobList []*JobT) error {
...
-		}
-		jd.writeChannel <- writeJobRequest
-		waitTimeStat.End()
-		errMap := <-respCh
-		return errMap
-	} else {
+	command := func() interface{} {
 		return jd.storeWithRetryEach(jobList)
 	}
+	res, _ := jd.executeDbRequest(newWriteDbRequest("store", nil, command)).(map[uuid.UUID]string)
```
Comment (chandumlg) [inline jobsdb/jobsdb.go:3454]: store -> store_retry_each

**Review by chandumlg — CHANGES_REQUESTED** [review]
Small stat name change required.

Nice refactoring 👏

File: jobsdb/jobsdb.go (hunk 2 above)
Comment (atzoum) [reply jobsdb/jobsdb.go:3454]: thanks, I missed it

**Comment by Sidddddarth** [comment]
same formatting(both op or `readReq`, `writeReq`) for read and write channels
```
for op := range jd.readChannel {
	op.response <- op.execute()
}
```

```
for writeReq := range jd.writeChannel {
        writeReq.response <- writeReq.execute()
}
```
LGTM

**Review by Sidddddarth — APPROVED** [review]

**Review by chandumlg — APPROVED** [review]

**Pull request merged**